### DISCUSSION TITLE
Prompt Caching?

### DISCUSSION THREAD

**Discussion by brandonh-msft** [microsoft/semantic-kernel#9444] [Q&A]
Prompt caching for [AOAI](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/prompt-caching) and [OAI](https://openai.com/index/api-prompt-caching/) has been released, but has [some requirements](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/prompt-caching#what-is-cached) around how to make it most effective, specifically, _the order_ in which things are serialized.

Is SK already serializing objects in this manner so token caching will get maximum hits? If not, is there work on the backlog to do it and when could we expect it (for each lang)?

**Comment by evchaki** [comment]
@brandonh-msft - thanks for brining this up.  This looks like it is happening on the model end so you should be getting this already.  Cosmos DB and Redis also supports Semantic Caching if you want to check that out also.

You can also do Semantic Caching with filters also here - https://github.com/microsoft/semantic-kernel/blob/2065883710150522fccb3cc1f0260faad74705e6/dotnet/samples/Concepts/Caching/SemanticCachingWithFilters.cs#L10

**Reply by brandonh-msft** [reply]
> @brandonh-msft - thanks for brining this up. This looks like it is happening on the model end so you should be getting this already. 

Right, but the _request_ has to be formatted propertly in order for it to have maximal effect (see the "some requirements" link), specifically, the oft-repeated part of the body should be at the top (e.g. tool definitions, chat history) as the caching stops "permanently" (for the request) as soon as it hits the first differing token.

Hence my question; when SK sends the payload to the model endpoint, are we creating it in such a way as to maximize prompt caching?

**Reply by evchaki** [reply]
We are sending the request the same way each time (unless you are dynamically adding plugins - those could be in a different order depending on how you add them)

**Reply by brandonh-msft** [reply]
of course but _in_ the request, are we making sure to send the tools array first, then chat _history_, then new messages? This would maximize the effect of prompt caching.

**Reply by ManojG1978** [reply]
A related question would be - is it possible for SK to report Cached tokens in the trace messages when hooked up with Open Telemetry. It would be nice to get that value along with the info generated:

![image](https://github.com/user-attachments/assets/d8ceff46-d4e9-42fc-a358-a48f5b30a80d)

Basically, everything OpenAI returns (including cached tokens, like below):
![image](https://github.com/user-attachments/assets/3e2c9bb8-71a5-4750-8f17-b7ed9b43524b)

**Reply by evchaki** [reply]
@TaoChenOSU - can you see if we can pull in the additional info from OpenAI.

**Reply by evchaki** [reply]
@brandonh-msft - this will depend how many tokens you are using with your plugin description and the length of chat history.

**Reply by brandonh-msft** [reply]
i understand that, but it won't do _anything at all_ if we aren't structuring the request to take advantage of it. this is what I opened this discussion to ask.

From the responses thus far it seems like the straightforward answer is "no, we aren't" so I will probably transition this to an issue if/when have time to dissect the code & confirm. (unless one of the other participants here wants to do that 🙏🏻)

**Reply by brandonh-msft** [reply]
It looks like this behavior is controlled clear down in the OpenAI pkg for .NET

https://github.com/openai/openai-dotnet/blob/e4af16915b120799c34786cc6b05c0e25e632b02/src/Custom/Chat/ChatClient.cs#L118

https://github.com/openai/openai-dotnet/blob/c49dd7065215bc0d094c7f79ccd634a38f0d7b66/src/Generated/Models/ChatCompletionOptions.Serialization.cs#L627

and so it's on them to make sure this is being done. I will open discussion/issue accordingly.

**Comment by brandonh-msft** [comment]
Issue filed on OpenAI .NET SDK: https://github.com/openai/openai-dotnet/issues/281
